### PR TITLE
feat: add Outlook HTTP API validation

### PR DESCRIPTION
**Pull request by PsypherPunk**
- check the validity of Outlook/Office 365 email addresses via the method outlined [here](https://www.trustedsec.com/blog/achieving-passive-user-enumeration-with-onedrive/).
- run only via the `--outlook-use-api` flags (defaulting to `false`.)

relates #937

### CODE REVIEW

File: core/src/smtp/mod.rs [hunk 1]
```diff
@@ -69,6 +69,11 @@ pub async fn check_smtp(
 			.await
 			.map_err(|err| err.into());
 	}
+	if input.outlook_use_api && host_lowercase.ends_with(".outlook.com.") {
```
Comment (PsypherPunk) [inline core/src/smtp/mod.rs:72]: Hotmail addresses seem to resolve to `hotmail-com.olc.protection.outlook.com.` so they'll get caught by this clause. However, it will result in an error as the OneDrive trick doesn't work for that provider.

All Office365-backed domains, however, seem to resolve to `<domain>.mail.protection.outlook.com.` so other than specifically excluding the above `hotmail-com…` host, I'm not sure if there's an easy distinction.

File: core/src/smtp/hotmail.rs [hunk 2]
```diff
@@ -193,4 +251,12 @@ mod tests {
 		let f = f1.try_join(f2).await;
 		assert!(f.is_ok(), "{:?}", f);
 	}
+
+	#[test]
```
Comment (PsypherPunk) [inline core/src/smtp/hotmail.rs:255]: No tests for the `check_outlook_api()` function—any idea if there's a `someone@gmail.com` equivalent that could be used here?

File: core/src/smtp/mod.rs (hunk 1 above)
Comment (amaurym) [reply core/src/smtp/mod.rs:72]: I tested a couple of emails:
- @outlook.com -> `outlook-com.olc.protection.outlook.com.`
- @outlook.fr -> `eur.olc.protection.outlook.com.`
- @hotmail.com -> `hotmail-com.olc.protection.outlook.com.` 
- @hotmail.fr -> `eur.olc.protection.outlook.com.`
- @domain.tld -> `domain-tld.mail.protection.outlook.com.` (tested with 2 different companies)

So it seems to me that we can do:
```suggestion
	if input.outlook_use_api && host_lowercase.ends_with(".mail.protection.outlook.com.") {
```

**Comment by amaurym** [comment]
Hey @PsypherPunk, after some testing, I am a bit reluctant to merge this PR. I tested on some valid Outlook 365 emails, and the results are all incorrect: Reacher returns `invalid` when they are `valid`.

If you send me your email, I can send you a couple of emails to test against. You can find mine in the README.

**Comment by PsypherPunk** [comment]
My email's in the commit message if you want to forward those addresses to test: will be happy to give them a whirl see if there's something I can tweak.

**Comment by amaurym** [comment]
Done. Yeah, let's play a bit to see if we can make this work 💪

**Comment by PsypherPunk** [comment]
I've tried this with those sample emails and yep, it fails miserably.

However, I've tried this with another couple of domains—presumably ones where OneDrive use is more ubiquitous—and it's 100% accurate. So, as the article suggests, positive responses are accurate but false-negatives are possible (and seemingly likely.)

Is there any value in doing something like this?

```rust
if input.outlook_use_api && host_lowercase.ends_with(".mail.protection.outlook.com.") {
	if let Some(smtp_details) = hotmail::check_outlook_api(to_email, input)
		.await
		.map_err(|err| err.into()) {
		return smtp_details;
	}
}
```

Where `check_outlook_api()` would return `None` in the event of 404, allowing further checks to be carried out…? That way, it's not taking a failed response as an absolute.

**Comment by hckhanh** [comment]
Hi @amaurym and @PsypherPunk, Is there any news for this PR? Is there anything that I can help?

**Comment by PsypherPunk** [comment]
Have you given it a try, @hckhanh? I'd be curious to know how it's working for anyone else.

**Comment by amaurym** [comment]
> positive responses are accurate but false-negatives are possible

and

> Where check_outlook_api() would return None in the event of 404, allowing further checks to be carried out…?

I think this makes sense @PsypherPunk, let's do this. Can you please also add relevant comments detailing the drawbacks of this `check_outlook_api` approach?

**Comment by hckhanh** [comment]
Hi @PsypherPunk, I think the result is **really** good:

```json
{
  "input": "hckhanh@[redacted].com",
  "is_reachable": "safe",
  "misc": {
    "is_disposable": false,
    "is_role_account": false,
    "gravatar_url": null
  },
  "mx": {
    "accepts_mail": true,
    "records": [
      "[redacted]-com.mail.protection.outlook.com."
    ]
  },
  "smtp": {
    "can_connect_smtp": true,
    "has_full_inbox": false,
    "is_catch_all": false,
    "is_deliverable": true,
    "is_disabled": false
  },
  "syntax": {
    "address": "hckhanh@[redacted].com",
    "domain": "[redacted].com",
    "is_valid_syntax": true,
    "username": "hckhanh"
  }
}
```

Even the result is much better than the previous one (it was `invalid`)

**Comment by hckhanh** [comment]
I have another question, @PsypherPunk
Is there any support for iCloud through API?

**Comment by PsypherPunk** [comment]
Thanks for the output, @hckhanh.

Re. iCloud, I'm afraid I've no idea. The changes here were raised as part of #937 which are specific to Google and Outlook. There _might_ be something similar for iCloud, but I'm not aware of it.

File: cli/src/main.rs [hunk 3]
```diff
@@ -63,6 +63,11 @@ pub struct Cli {
 	#[clap(long, env, default_value = "false", parse(try_from_str))]
 	pub gmail_use_api: bool,
 
+	/// For Outlook/Office 365 email addresses, use Outlook's API instead of
```
Comment (amaurym) [inline cli/src/main.rs:66]: I have a question about the name "Outlook". What's the official name of their email subscription service for businesses? I always thought it was "Microsoft 365", but it seems "Office 365" also exists, or is the latter only referring to Word/Excel/PPT..?

I would propose to consolidate around one name, e.g.:
```suggestion
	/// For Microsoft 365 email addresses, use OneDrive's API instead of
```

File: cli/src/main.rs [hunk 4]
```diff
@@ -63,6 +63,11 @@ pub struct Cli {
...
+	/// connecting directly to their SMTP servers.
+	#[clap(long, env, default_value = "false", parse(try_from_str))]
+	pub outlook_use_api: bool,
```
Comment (amaurym) [inline cli/src/main.rs:69]:
```suggestion
	pub microsoft365_use_api: bool,
```

**Review by amaurym — COMMENTED** [review]
Code LGTM! I have one final comment about the public-facing API. Using `outlook_` may seem like this method would work for @outlook.com emails too, which is incorrect.

Would like to here opinions about:
- `microsoft365_`
- `office365_`

I'm not too well-versed in the MS ecosystem to know what terms people usually use.

**Comment by PsypherPunk** [comment]
Good shout, @amaurym: Wikipedia says [_Microsoft 365_](https://en.wikipedia.org/wiki/Microsoft_365): _"Microsoft 365, formerly Office 365…"_ so I'll go with that…?

File: core/src/smtp/hotmail.rs [hunk 5]
```diff
@@ -140,9 +151,65 @@ pub async fn check_password_recovery(
...
+/// for details on the underlying principles.
+///
+/// Note that a positive response from this function is (at present) considered
+/// a reliable indicator that an email-address is valid. However, a negative
+/// response is ambigious: the email address may or may not be valid but this
+/// cannot be determined by the method outlined here.
+pub async fn check_microsoft365_api(
+	to_email: &EmailAddress,
+	input: &CheckEmailInput,
+) -> Result<Option<SmtpDetails>, HotmailError> {
```
Comment (amaurym) [inline core/src/smtp/hotmail.rs:184]: I just tried this manually, and I believe there's a bug: if you run `cargo run -- --microsoft365-use-api=true <email>` with someone that hasn't opened their one drive, it returns a ReqwestError. Whereas I believe in https://github.com/reacherhq/check-if-email-exists/pull/1194#issuecomment-1271578587 we both meant that in that case the algorithm should just continue as normal.

Comment (amaurym) [reply core/src/smtp/hotmail.rs:184]: I'm inclined to simply do this:

```suggestion
) -> Result<SmtpDetails, HotmailError> {
```

and  below do:

```rust
	if input.microsoft365_use_api && host_lowercase.ends_with(".mail.protection.outlook.com.") {
		// Continue in the event of an ambiguous result.
		if let Ok(smtp_details) = hotmail::check_microsoft365_api(to_email, input).await {
			return Ok(smtp_details);
		}
	}
```

Comment (PsypherPunk) [reply core/src/smtp/hotmail.rs:184]: Ah, I think I know what's happening:

1. if the API returns a 403, it returns a positive response (`Ok(Some(smtp_details))`.)
1. if the API returns a 404, it returns a negative response (`Ok(None)`.)
1. if the API fails because the OneDrive domain doesn't exist (i.e. the company in question just doesn't use OneDrive), it raises an error (`Err(HotmailError(ReqwestError))`.)
1. if the API fails for some other reason (e.g. a transient network issue), it raises an error (`Err(HotmailError(ReqwestError))`.)

The problem is that those last two are indistinguishable, although 3. should result in some indication that it's ambiguous, like 1. and 2.

However, I can't of a way to make that distinction so I think you're right: I'll aim to allow it continue in the event of _any_ error (but might add an additional log which might allow things to be narrowed down in the future.)

Comment (PsypherPunk) [reply core/src/smtp/hotmail.rs:184]: @amaurym, I've pushed up a change which resolves the issues (I think; I've tested the scenarios I'm aware of any they either pass or continue to SMTP checks) but leaves the `Result<Option<SmtpDetails>, HotmailError>` signature in place.

In order to have `check_microsoft365_api()` return a `Result<SmtpDetails, HotmailError>`, I'd have to change the "successful" 404 response into a `HotmailError`. That's easy enough but adding another `HotmailError` type (e.g. `HotmailError::AmbiguousResponse`) didn't seem quite right.

Basically:

- if the HTTP request succeeds and we see a 403, we return the result.
- if the HTTP request succeeds and we see a 404, we continue to SMTP checks.
- if the HTTP request fails, we log the error and continue to SMTP checks.

**Review by amaurym — APPROVED** [review]
This LGTM, thanks @PsypherPunk!

**Comment by hckhanh** [comment]
Hi @amaurym, @PsypherPunk 
I try to switch to master branch to check this feature, but I get this bug when I build the project

```
error[E0433]: failed to resolve: use of undeclared crate or module `hotmail`
  --> C:\Users\hoang\.cargo\git\checkouts\check-if-email-exists-c4445d2e2cd9a1bc\5d3c49f\core\src\smtp\mod.rs:73:9
   |
73 |         match hotmail::check_microsoft365_api(to_email, input).await {
   |               ^^^^^^^ use of undeclared crate or module `hotmail`

For more information about this error, try `rustc --explain E0433`.
error: could not compile `check-if-email-exists` due to previous error
warning: build failed, waiting for other jobs to finish...
Internal Error: Command failed: cargo build
```
